pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -88,6 +91,130 @@ func TestCreateAndDeleteOfAffinityAssistant(t *testing.T) {
+func TestCreateAffinityAssistantWhenNodeIsCordoned(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	c := Reconciler{
+		KubeClientSet: fakek8s.NewSimpleClientset(),
+		Images:        pipeline.Images{},
+	}
+
+	workspaceName := "testws"
+	pipelineRunName := "pipelinerun-1"
+	testPipelineRun := &v1beta1.PipelineRun{
+		TypeMeta: metav1.TypeMeta{Kind: "PipelineRun"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pipelineRunName,
+		},
+		Spec: v1beta1.PipelineRunSpec{
+			Workspaces: []v1beta1.WorkspaceBinding{{
+				Name: workspaceName,
+				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+					ClaimName: "myclaim",
+				},
+			}},
+		},
+	}
+
+	// create affinity assistant in a cluster
+	err := c.createAffinityAssistants(ctx, testPipelineRun.Spec.Workspaces, testPipelineRun, testPipelineRun.Namespace)
+	if err != nil {
+		t.Errorf("unexpected error from createAffinityAssistants: %v", err)
+	}
+
+	// get the affinity assistant just created
+	expectedAffinityAssistantName := getAffinityAssistantName(workspaceName, testPipelineRun.Name)
+
+	// assume the necessary pod is created, update the affinity assistant with the number of pods in readyReplicas
+	var replaceReplicasPatchBytes []byte
+	replaceReplicasPatchBytes, err = json.Marshal([]jsonpatch.JsonPatchOperation{{
+		Operation: "replace",
+		Path:      "/status/readyReplicas",
+		Value:     int32(1),
+	}})
+	_, err = c.KubeClientSet.AppsV1().StatefulSets(testPipelineRun.Namespace).Patch(ctx, expectedAffinityAssistantName, types.JSONPatchType, replaceReplicasPatchBytes, metav1.PatchOptions{})
+	if err != nil {
+		t.Errorf("unexpected error updating StatefulSet: %v", err)
+	}
+
+	// register cordoned node to the cluster
+	c.KubeClientSet.CoreV1().Nodes().Create(ctx, &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "soon-to-be-cordoned-node",
+		},
+		Spec: corev1.NodeSpec{
+			Unschedulable: true,
+		},
+	}, metav1.CreateOptions{})
+
+	// now, createAffinityAssistants should be able to catch the missing affinity assistant pod with an existing StatefulSet and unschedulable node
+	err = c.createAffinityAssistants(ctx, testPipelineRun.Spec.Workspaces, testPipelineRun, testPipelineRun.Namespace)
+	if err == nil {
+		t.Error("expected error from createAffinityAssistants, but no error was reported")
+	}
+
+	p := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: expectedAffinityAssistantName + "-0",
+			Labels: map[string]string{
+				workspace.LabelComponent:             workspace.ComponentNameAffinityAssistant,
+				workspace.LabelInstance:              expectedAffinityAssistantName,
+				"statefulset.kubernetes.io/pod-name": expectedAffinityAssistantName + "-0",
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: "apps/v1",
+				Kind:       "StatefulSet",
+				Name:       expectedAffinityAssistantName,
+			}},
+		},
+	}
+
+	// create an affinity-assistant pod with necessary labels
```

Comment:
Can you separate this test case into separate test cases? It seems like we want to test:
- creating a new statefulset when one doesn't already exist
- doing nothing when a statefulset already exists and nodes are schedulable
- deleting pods when the statefulset already exists and the nodes are not schedulable
I think it would make sense for these to be separate tests.

Reply:
This test function tests three separate test cases:

* Test 1: createAffinityAssistants must return an error when the StatefulSet has the readyReplicas set to 1 without any pod with the necessary label created

* Test 2: createAffinityAssistants must delete an affinity assistant pod since the node on which its scheduled is marked as unschedulable

* Test 3: createAffinityAssistants must return an error when the affinity assistant pod has a nodeName which does not exist

I have added explicit comments in one single function. Let me know if you still think, we should separate them out. I avoided creating three separate functions since its testing the same code block.

Reply:
I still think it would make sense to split these out into separate test cases. If we have to have some duplication in tests that's OK IMO.

Reply:
thanks @lbernick Sounds good! I prefer it to be part of a single test function and will keep it as is for now. I will let the additional reviewer chime in and change it as needed.